pkg/plugin/plugin_v2.go

```diff
@@ -0,0 +1,205 @@
+	if !recent {
+		_, err = p.Encrypt(context.Background(), &pb.EncryptRequest{Plaintext: []byte("foo")})
+		p.healthCheck.recordErr(err)
+		if err != nil {
+			zap.L().Warn("health check failed", zap.Error(err))
+		}
+		return err
+	}
+	if err != nil {
+		zap.L().Warn("cached health check failed", zap.Error(err))
+	} else {
+		zap.L().Debug("health check success")
+	}
+	return err
+}
+
+func (p *V2Plugin) Live() error {
+	if err := p.Health(); err != nil && kmsplugin.ParseError(err) != kmsplugin.KMSErrorTypeUserInduced {
+		return err
+	}
+	return nil
+}
+
+// Status returns the V2Plugin server status
+func (p *V2Plugin) Status(ctx context.Context, request *pb.StatusRequest) (*pb.StatusResponse, error) {
+	status := "ok"
+	if p.Health() != nil {
+		status = "err"
+	}
+	return &pb.StatusResponse{
+		Version: "v2beta1",
+		Healthz: status,
+		KeyId:   p.keyID,
+	}, nil
+}
+
+// Encrypt executes the encryption operation using AWS KMS
+func (p *V2Plugin) Encrypt(ctx context.Context, request *pb.EncryptRequest) (*pb.EncryptResponse, error) {
+	zap.L().Debug("starting encrypt operation")
+
+	startTime := time.Now()
+	input := &kms.EncryptInput{
+		Plaintext: request.Plaintext,
+		KeyId:     aws.String(p.keyID),
+	}
+	if len(p.encryptionCtx) > 0 {
+		zap.L().Debug("configuring encryption context", zap.String("ctx", fmt.Sprintf("%v", p.encryptionCtx)))
+		input.EncryptionContext = p.encryptionCtx
+	}
+
+	result, err := p.svc.Encrypt(input)
+	if err != nil {
+		select {
+		case p.healthCheck.healthCheckErrc <- err:
+		default:
+		}
+		zap.L().Error("request to encrypt failed", zap.String("error-type", kmsplugin.ParseError(err).String()), zap.Error(err))
+		failLabel := kmsplugin.GetStatusLabel(err)
+		kmsLatencyMetric.WithLabelValues(p.keyID, failLabel, kmsplugin.OperationEncrypt, GRPC_V2).Observe(kmsplugin.GetMillisecondsSince(startTime))
+		kmsOperationCounter.WithLabelValues(p.keyID, failLabel, kmsplugin.OperationEncrypt, GRPC_V2).Inc()
+		return nil, fmt.Errorf("failed to encrypt %w", err)
+	}
+
+	zap.L().Debug("encrypt operation successful")
+	kmsLatencyMetric.WithLabelValues(p.keyID, kmsplugin.StatusSuccess, kmsplugin.OperationEncrypt, GRPC_V2).Observe(kmsplugin.GetMillisecondsSince(startTime))
+	kmsOperationCounter.WithLabelValues(p.keyID, kmsplugin.StatusSuccess, kmsplugin.OperationEncrypt, GRPC_V2).Inc()
+	return &pb.EncryptResponse{
+		Ciphertext: append([]byte(kmsplugin.StorageVersion), result.CiphertextBlob...),
+		KeyId:      p.keyID,
+	}, nil
+}
+
+// Decrypt executes the decrypt operation using AWS KMS
+func (p *V2Plugin) Decrypt(ctx context.Context, request *pb.DecryptRequest) (*pb.DecryptResponse, error) {
+	zap.L().Debug("starting decrypt operation")
+
+	startTime := time.Now()
+	if string(request.Ciphertext[0]) == kmsplugin.StorageVersion {
+		request.Ciphertext = request.Ciphertext[1:]
+	} else {
+		// enforce the kmsplugin.StorageVersion in v2
+		return nil, fmt.Errorf("version in Ciphertext doesn't match kmsplugin")
+	}
```

Comment:
You need to use the `kmsplugin.StorageVersionV2`

Maybe a switch would be nice here to make future refactors simpler?

And returning the discovered version would be useful for debugging

```suggestion
	// TODO: check Ciphertext has length > 2
	switch version := request.Ciphertext[0]; version { 
	case kmsplugin.StorageVersionV2:
		request.Ciphertext = request.Ciphertext[1:]
	default:
		// enforce the kmsplugin.StorageVersion in v2
		return nil, fmt.Errorf("version %d in Ciphertext doesn't match kmsplugin", version)
	}
```